helix-core/src/syntax.rs

```diff
@@ -371,6 +371,9 @@ pub struct LanguageServerConfiguration {
     pub config: Option<serde_json::Value>,
     #[serde(default = "default_timeout")]
     pub timeout: u64,
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    pub required_root_patterns: Vec<String>,
```

Comment:
Let's do the compilation into a globset at config parsing time instead, adding `globset` to `helix-core` (we want to do this anyways for #8006):

```rust
pub struct LanguageServerConfiguration {
    ..
    #[serde(
        default,
        skip_serializing,
        deserialize_with = "deserialize_required_root_patterns"
    )]
    pub required_root_patterns: Option<GlobSet>,
}
```

```rust
fn deserialize_required_root_patterns<'de, D>(deserializer: D) -> Result<Option<GlobSet>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let patterns = Vec::<String>::deserialize(deserializer)?;
    if patterns.is_empty() {
        return Ok(None);
    }
    let mut builder = globset::GlobSetBuilder::new();
    for pattern in patterns {
        let glob = globset::Glob::new(&pattern).map_err(serde::de::Error::custom)?;
        builder.add(glob);
    }
    builder.build().map(Some).map_err(serde::de::Error::custom)
}
```

Then the check in `start_client` becomes really simple:

```rust
if let Some(globset) = &ls_config.required_root_patterns {
    if !root_path
        .read_dir()?
        .flatten()
        .map(|entry| entry.file_name())
        .any(|entry| globset.is_match(entry))
    {
        return Ok(None);
    }
}
```

Reply:
```rs
let glob = globset::Glob::new(&pattern).map_err(serde::de::Error::custom)?;
```
Would this drop the entire required roots list if any glob pattern couldn't be built? My intention was to log an error on bad patterns, but still check the ones that could be built.

Reply:
The whole language configuration parse would fail, so you would see the Bad config error on startup: https://github.com/helix-editor/helix/blob/d570c29ce37ffbb46a9c49708c31dfd81daa27cf/helix-term/src/main.rs#L149-L150

I would prefer to fail early and loudly in this case since it can be confusing when the config successfully parses and the editor starts but not everything you wanted to configure works

Reply:
Ah ok, makes sense then.